File: playwright-java/src/main/java/testtracks/diagnostics/UiEvidence.java
package testtracks.diagnostics;

import com.microsoft.playwright.Page;
import java.util.ArrayList;
import java.util.List;

public final class UiEvidence {
  private final List<String> consoleMessages = new ArrayList<>();
  private final List<String> pageErrors = new ArrayList<>();

  public List<String> consoleMessages() {
    return List.copyOf(consoleMessages);
  }

  public List<String> pageErrors() {
    return List.copyOf(pageErrors);
  }

  public void attachTo(Page page) {
    page.onConsoleMessage(
        message ->
            consoleMessages.add(SecretRedactor.redact(message.type() + ": " + message.text())));
    page.onPageError(error -> pageErrors.add(SecretRedactor.redact(error)));
  }
}

File: playwright-csharp/src/Diagnostics/UiEvidence.cs
using Microsoft.Playwright;

namespace TestTracks.Playwright.CSharp.Diagnostics;

/// <summary>
/// Captures browser console messages and page errors for a UI scenario.
/// </summary>
public sealed class UiEvidence
{
    private readonly List<string> _consoleMessages = [];
    private readonly List<string> _pageErrors = [];

    public IReadOnlyList<string> ConsoleMessages => _consoleMessages;
    public IReadOnlyList<string> PageErrors => _pageErrors;

    /// <summary>
    /// Subscribes to Playwright page events so UI evidence is captured as the scenario runs.
    /// </summary>
    public void AttachTo(IPage page)
    {
        page.Console += (_, message) =>
            _consoleMessages.Add(SecretRedactor.Redact($"{message.Type}: {message.Text}"));

        page.PageError += (_, error) =>
            _pageErrors.Add(SecretRedactor.Redact(error));
    }
}
